### ISSUE TITLE
Sprites visualization error on IE9

### ISSUE THREAD

**Issue by potomak** [craftyjs/Crafty#549]
I'm rendering a 2D component but it doesn't show in IE9.

The component is named `Monster3D`, this is the source code:

``` javascript
Crafty.c('Monster3D', {
  init: function() {
    this.requires('Actor3D, spriteMonster, SpriteAnimation')
      .attr({three: {x: 0, y: 0, z: -100, w: 40, h: 40}})
      .animate('MonsterMoving', 0, 0, 18);

    this.animate('MonsterMoving', 100, -1);
  }
});
```

The `Actor3D` component is a custom component composed by `Canvas` and `3D`, the latest is a simple custom component to render canvas components in a 3D space.

``` javascript
Crafty.c('Actor3D', {
  init: function() {
    this.requires('3D, Canvas');
  },

  at: function(x, y, z) {
    this.three.x = x;
    this.three.y = y;
    this.three.z = z;

    return this;
  }
});
```

I put 3x5 `Monster3D` components by running:

``` javascript
Crafty.e('Monster3D').at(-200, 0, 300);
Crafty.e('Monster3D').at(-100, 0, 300);
Crafty.e('Monster3D').at(   0, 0, 300);
Crafty.e('Monster3D').at( 100, 0, 300);
Crafty.e('Monster3D').at( 200, 0, 300);
Crafty.e('Monster3D').at(-200, 0, 200);
Crafty.e('Monster3D').at(-100, 0, 200);
Crafty.e('Monster3D').at(   0, 0, 200);
Crafty.e('Monster3D').at( 100, 0, 200);
Crafty.e('Monster3D').at( 200, 0, 200);
Crafty.e('Monster3D').at(-200, 0, 100);
Crafty.e('Monster3D').at(-100, 0, 100);
Crafty.e('Monster3D').at(   0, 0, 100);
Crafty.e('Monster3D').at( 100, 0, 100);
Crafty.e('Monster3D').at( 200, 0, 100);
```

At http://popinator.fungostudios.com/error/ there's the test script.

Chrome (Ubuntu 13.04):
![screenshot from 2013-09-18 15 21 05](https://f.cloud.github.com/assets/153842/1165099/a9b6221c-2066-11e3-9a76-3519ff6800ed.png)

IE9 (Windows 7):
![popinator_error](https://f.cloud.github.com/assets/153842/1165100/aff0ebda-2066-11e3-8599-d754c161f517.png)

**Note: if I try to run the latest snippet of code inside a timeout function it renders the sprites.**

**Comment by starwed** [comment]
Hmm, it's impossible for me to really test this, since I have no devices which run IE.  

If it works in a timeout function, could it just be that the images haven't loaded yet?

**Comment by potomak** [comment]
@starwed thanks for your reply!

Well, it happens that without timeout the first load is ok (monsters are showed), but reloads fail, so I don't think it's matter of image loading.

It's a shame that we can't test IE easily.

**Comment by starwed** [comment]
Do you get any error messages?

Can you tell if the missing entities are actually trying to draw themselves?  You should be able to test this with something like `e.bind("Draw", function(){console.log("I'm getting rendered!")})`.   (Except with whatever IE9 uses to log with.)

You can also try adding the "VisibleMBR" component if you're using a recent enough version, that should draw a rectangle representing the MBR on a separate debugging canvas.

**Comment by potomak** [comment]
No errors. Everything looks fine.

I don't have a Windows machine with me right now, tomorrow at work I'll try to log the `Draw` event, thanks for the support.

**Comment by sbyrnes321** [comment]
I can reproduce the error in IE10.

There is a lot going on in your code, I can't figure out what's the core of the problem.

When I looked into image-loading code a while ago, I reached the conclusion that the code had bugs that could cause race conditions:

https://groups.google.com/forum/#!msg/craftyjs/xiHBnwl-aHQ/1-5mSp_QG0QJ

But I never followed up.

Does it help if you use `Crafty.load(...)` (http://craftyjs.com/api/Crafty-loader.html) and let it finish before running the `Crafty.sprite(...)` and `Crafty.e(...)` commands?

**Comment by potomak** [comment]
@starwed I updated the code at http://popinator.fungostudios.com/error/, I can see console.logs either from the ball and from monsters with chrome, but with IE (even IE 10) I can see only ball consoles.

Example chrome log:

```
I'm a monster at (45.87255270355078, 294.86620163750763) and I'm getting rendered! milka-toss.js:136
I'm a monster at (115.93048329284588, 294.86620163750763) and I'm getting rendered! milka-toss.js:136
I'm a monster at (185.98841388214097, 294.86620163750763) and I'm getting rendered! milka-toss.js:136
I'm a monster at (256.0463444714361, 294.86620163750763) and I'm getting rendered! milka-toss.js:136
I'm a monster at (326.1042750607312, 294.86620163750763) and I'm getting rendered! milka-toss.js:136
I'm a monster at (23.970046908321418, 356.81347352886974) and I'm getting rendered! milka-toss.js:136
I'm a monster at (103.9836619499935, 356.81347352886974) and I'm getting rendered! milka-toss.js:136
I'm a monster at (183.9972769916656, 356.81347352886974) and I'm getting rendered! milka-toss.js:136
I'm a monster at (264.01089203333765, 356.81347352886974) and I'm getting rendered! milka-toss.js:136
I'm a monster at (344.02450707500975, 356.81347352886974) and I'm getting rendered! milka-toss.js:136
I'm a monster at (-5.188574936777847, 439.2833558475044) and I'm getting rendered! milka-toss.js:136
I'm a monster at (88.07895912539391, 439.2833558475044) and I'm getting rendered! milka-toss.js:136
I'm a monster at (181.34649318756564, 439.2833558475044) and I'm getting rendered! milka-toss.js:136
I'm a monster at (274.61402724973743, 439.2833558475044) and I'm getting rendered! milka-toss.js:136
I'm a monster at (367.8815613119092, 439.2833558475044) and I'm getting rendered! milka-toss.js:136
I'm the ball at (223.0854238764454, 593.3497953506038) and I'm getting rendered! milka-toss.js:258
```

Example IE log:

```
I'm the ball at (223.0854238764454, 593.3497953506038) and I'm getting rendered! milka-toss.js:258
```

Note: I'm using CraftyJS 0.5.4 but I can't find the VisibleMBR component!

**Comment by potomak** [comment]
@sbyrnes321 yes, I'm using a loader, and yes, I load the `Game` only after assets are loaded and sprite are created.

On page load I execute  `Game.start`:

``` javascript
Game = {
  width: 400,
  height: 600,
  gravity: 0.5,

  start: function() {
    Crafty.init(Game.width, Game.height);
    Crafty.three.camera.y = -100;
    Crafty.background('url(assets/background.png)');

    Crafty.scene('Loading');
  },

  textCss: {
    'font-size': '34px',
    'font-family': 'Arial',
    'text-align': 'center'
  }
};
```

`Game.start` loads the `Loading` scene which calls `Crafty.load` and in its callback creates sprite objects and loads the `Game` scene, where everything else happens and where elements are placed in scene.

``` javascript
Crafty.scene('Loading', function() {
  var assets = [
    'assets/ball.png',
    'assets/monster.png',
    'assets/board_room_applause.mp3',
    'assets/board_room_applause.ogg',
    'assets/board_room_applause.aac'
  ];

  Crafty.load(assets, function() {
    Crafty.sprite(50, 'assets/ball.png', {
      spritePopcorn: [0, 0],
    }, 0, 2);

    Crafty.sprite(40, 'assets/monster.png', {
      spriteMonster: [0, 0],
    }, 0, 2);

    Crafty.audio.add({
      applause: [
        'assets/board_room_applause.mp3',
        'assets/board_room_applause.ogg',
        'assets/board_room_applause.aac'
      ]
    });

    Crafty.scene('Game');
  });
});
```

Am I doing something wrong or do you think it's a IE issue?

**Comment by starwed** [comment]
Sorry, the VisibleMBR component exists in the 0.6 alphas -- not in 0.5.4.

Your loading setup seems like it _should_ be ok to me.

The next basic debugging step would be to add logging to check your assumptions -- make sure the entities are being created in the first place, for instance.

You could check that by looking at `Crafty("Monster3D").length`, or by adding logging code directly into the init functions of the components.  Also grab a single monster entity and log the contents of `entity.__c` to check which components are present.

**Comment by potomak** [comment]
Yes, they are there, but they are not visualized.

From the console:

```
>> Crafty("Monster3D").length 
15
>> Crafty(Crafty("Monster3D")[0]).__c 
{
  Monster3D : true,
  Actor3D : true,
  3D : true,
  2D : true,
  Canvas : true,
  spriteMonster : true,
  Sprite : true,
  SpriteAnimation : true,
  obj : true
}
```

Now I'd like to try version 0.6 of the framework, is there a transition guide?

**Comment by potomak** [comment]
I tried the script with the develop branch of crafty and it seems to have the same problem.

If I put a timeout around component creation (`Crafty.e('Monster3D').at(..., ..., ...);`) it works well.

**Comment by starwed** [comment]
So it does sound like there's some problem with the image loading; perhaps IE is triggering the image's `onload` before it's actually loaded?  Or perhaps one of the other assets is triggering onload twice.  (The load function runs the callback once as many valid assets have loaded as were requested.)

You might want to add an onprogress callback to the load function, to see if there's anything weird about how the assets are loading.  Something like

``` javascript
onProgress = function(data){ 
    console.log("loaded " + data.src + ", asset " + data.loaded + " of " + data.total);
}
```

And then (with your existing `onComplete`)

``` javascript
Crafty.load(assets, onComplete, onProgress);
```

**Comment by potomak** [comment]
@starwed @sbyrnes321 I found the problem!

On https://github.com/craftyjs/Crafty/blob/develop/src/loader.js#L204 @starwed added the line

``` javascript
obj.src = ""; // workaround for webkit bug
```

before the actual loading of the asset.

This workaround that should fix a webkit bug (see 1685064) triggers the `onerror` callback on IE, `onerror` increases the number of loaded assets (see https://github.com/craftyjs/Crafty/blob/develop/src/loader.js#L175) and consequently triggers the `oncomplete` callback before assets are loaded.

The reason is that `imageObject.src = ""` triggers the error event on IE while it doesn't on webkit.

Example code (http://jsfiddle.net/CcuH9/1/)

``` javascript
var image = new Image();

image.onload = function(e) {
  console.log("Load", e);
};

image.onerror = function(e) {
  console.log("Error", e);
};

image.src = "";
image.src = "http://placehold.it/100x100.png";
```

To fix this error I suggest to run the webkit workaround only if the current browser is a webkit based browser, what do you think about it?

**Comment by potomak** [comment]
Closed by #562
